dash/include/dash/pattern/internal/PatternArguments.h

```diff
@@ -121,72 +104,160 @@ class PatternArguments {
+   *                  distribution of the pattern parsed so far.
+   *                  Set to -1 if a \c TeamSpec is encountered.
+   */
+  template<
+    int ArgcSize,
+    int ArgcDist,
+    int ArgcTeam,
+    int ArgcTeamSpec,
+    typename ... Args>
+  void parse(SizeType extent, Args && ... args) {
+    static_assert(ArgcSize >= 0, "Cannot mix size and SizeSpec definition"
+        "in variadic pattern constructor!");
     DASH_LOG_TRACE("PatternArguments.check(extent)", extent);
-    _argc_size++;
-    _sizespec.resize(count, extent);
+    static_assert(ArgcSize < NumDimensions, "Number of size specifier exceeds"
+        "the number of dimensions in variadic pattern constructor!");
+    DASH_LOG_TRACE("PatternArguments.check(extent)", extent);
+    _sizespec.resize(ArgcSize, extent);
+    parse<ArgcSize+1, ArgcDist, ArgcTeam,
+      ArgcTeamSpec>(std::forward<Args>(args)...);
+  }
+
+  /*
+   * Match for one optional parameter specifying the size (extents).
+   */
+  template<
+    int ArgcSize,
+    int ArgcDist,
+    int ArgcTeam,
+    int ArgcTeamSpec,
+    typename ... Args>
+  void parse(const SizeSpec_t & sizeSpec, Args && ... args) {
+    static_assert(ArgcSize == 0, "Cannot mix size and SizeSpec definition"
+        "in variadic pattern constructor!");
+    DASH_LOG_TRACE("PatternArguments.check(sizeSpec)");
+    _sizespec = sizeSpec;
+    parse<-1, ArgcDist, ArgcTeam,
+      ArgcTeamSpec>(std::forward<Args>(args)...);
   }
-  /// BlockPattern matching for up to \c NumDimensions optional 
-  /// parameters specifying the distribution pattern.
-  template<int count>
-  void check(const TeamSpec_t & teamSpec) {
+
+  /*
+   * Match for \c TeamSpec describing the distribution among the units
+   * in the team.
+   */
+  template<
+    int ArgcSize,
+    int ArgcDist,
+    int ArgcTeam,
+    int ArgcTeamSpec,
+    typename ... Args>
+  void parse(const TeamSpec_t & teamSpec, Args && ... args) {
+    static_assert(ArgcTeamSpec == 0, "Cannot specify TeamSpec twice in "
+        "variadic pattern constructor!");
     DASH_LOG_TRACE("PatternArguments.check(teamSpec)");
-    _argc_team++;
     _teamspec   = teamSpec;
+    parse<ArgcSize, ArgcDist, ArgcTeam, 1>(std::forward<Args>(args)...);
   }
-  /// BlockPattern matching for one optional parameter specifying the 
-  /// team.
-  template<int count>
-  void check(dash::Team & team) {
+
+  /*
+   * Match for one optional parameter specifying the team.
+   */
+  template<
+    int ArgcSize,
+    int ArgcDist,
+    int ArgcTeam,
+    int ArgcTeamSpec,
+    typename ... Args>
+  void parse(dash::Team & team, Args && ... args) {
+    static_assert(!(ArgcTeam > 0),
+        "Cannot specify Team twice in variadic "
+        "pattern constructor!");
     DASH_LOG_TRACE("PatternArguments.check(team)");
-    if (_argc_team == 0) {
-      _team     = &team;
-      _teamspec = TeamSpec_t(_distspec, team);
-    }
+    _team     = &team;
+    _teamspec = TeamSpec_t(team);
```

Comment:
Possibly overwrites previously set TeamSpec.
Should already be handled by check in line [255](https://github.com/dash-project/dash/pull/356/files#diff-703bc11c993d6436737bb7664d6b84a0R255).